src/tech/v3/dataset/impl/column_index_structure.clj

```diff
@@ -0,0 +1,85 @@
+(ns tech.v3.dataset.impl.column-index-structure
+  (:import [java.util TreeMap LinkedHashMap]
+           [tech.v3.datatype ListPersistentVector])
+  (:require [tech.v3.protocols.column :as col-proto]
+            [tech.v3.dataset.impl.column-base :refer [column-datatype-categorical?]]
+
+            [tech.v3.datatype :refer [elemwise-datatype clone ->buffer]]
+            [tech.v3.datatype.argops :refer [arggroup]]
+            [tech.v3.datatype.casting :refer [datatype->object-class]]
+            [clojure.set :refer [difference]]))
+
+
+(extend-type TreeMap
+  col-proto/PIndexStructure
+  (select-from-index [index-structure mode selection-spec]
+    (case mode
+      :pick
+      (let [^TreeMap new-index-structure (.clone ^TreeMap index-structure)
+            s (difference (set (.keySet new-index-structure)) (set selection-spec))]
+        (doseq [k s]
+          (.remove new-index-structure k))
+        new-index-structure)
+      :slice
+      (let [{from            :from
+             from-inclusive? :from-inclusive?
+             to              :to
+             to-inclusive?   :to-inclusive?} selection-spec]
+        (.subMap ^TreeMap index-structure
+                 from
+                 (if (nil? from-inclusive?) true from-inclusive?)
+                 to
+                 (if (nil? to-inclusive?) true to-inclusive?))))))
+
+
+(extend-type LinkedHashMap
+  col-proto/PIndexStructure
+  (select-from-index [index-structure mode selection-spec]
+    (case mode
+      :pick
+      (let [^LinkedHashMap new-index-structure (.clone ^LinkedHashMap index-structure)
+            s (difference (set (.keySet new-index-structure)) (set selection-spec))]
+        (doseq [k s]
+          (.remove new-index-structure k))
+        new-index-structure))))
+
```

Comment:
These select-keys operations on java.util.Map class operations I think should be refactored into a common method.